src/Api/Serializer/AbstractSerializer.php

```diff
@@ -153,27 +140,21 @@ public function getRelationship($model, $name)
      */
     protected function getCustomRelationship($model, $name)
     {
-        // Deprecated in beta 15, removed in beta 16
-        $relationship = static::$dispatcher->until(
-            new GetApiRelationship($this, $name, $model)
-        );
-
         foreach (array_merge([static::class], class_parents($this)) as $class) {
             $callback = Arr::get(static::$customRelations, "$class.$name");
 
             if (is_callable($callback)) {
                 $relationship = $callback($this, $model);
-                break;
-            }
-        }
 
-        if ($relationship && ! ($relationship instanceof Relationship)) {
-            throw new LogicException(
-                'GetApiRelationship handler must return an instance of '.Relationship::class
-            );
-        }
+                if (! ($relationship instanceof Relationship)) {
+                    throw new LogicException(
+                        'GetApiRelationship handler must return an instance of '.Relationship::class
+                    );
+                }
```

Comment:
hmm, I'm not sure here.

The method used to return `Relationship|null`, but with this change, it's guaranteed to always return Relationship or error out, meaning callbacks which used to be able to return null and work fine, will error out.

I can't think of any real examples why a relationship callback would return null, not have I looked into it, but this is definitely a breaking change.

Reply:
Ah nevermind, it wouldn't get passed the is_callable test

Reply:
Actually, you bring up a really good point. That IS actually a valid use case, and in one of my own extensions no less (https://github.com/askvortsov1/flarum-moderator-warnings/blob/master/extend.php#L65-L72)! My goodness is this PR a mess...